### ISSUE TITLE
Request to Revamp twinBASIC Search Functionality to Match VB6 Standards for Simplicity and Efficiency

### ISSUE THREAD

**Issue by antoninononooono** [twinbasic/twinbasic#1896]
**Problem Description TLDR version**: Please make the search function in TwinBasic **exactely** like in VB6.
 Task completed.

**Problem Description:** In my experience, utilizing the .NET search function has proven to be problematic due to what appears to be fundamental design issues. Unfortunately, twinBASIC replicates these design shortcomings in its search function, and in some ways exacerbates them. The existing implementation of a global search seems unnecessary and overcomplicated. What is needed instead is a streamlined, efficient search function—akin to what was provided in VB6—without extraneous features such as byte counts or the last edit dates of the strings found. The focus should be on a singular, effective search tool that prioritizes simplicity and user experience.

**Proposed Solution:** I suggest a revision of the search functionality in twinBASIC to closely emulate the search capabilities of VB6. The search should be intuitive, direct, and robust, without requiring toggles between local and global search contexts. Users expect and depend on a search feature that reliably finds what they're looking for quickly and efficiently, as was the case with VB6.

**Rationale:** The simplicity and effectiveness of VB6’s search function set a standard that many developers appreciated. The current approach in environments like .NET, which often feels cumbersome and unreliable by comparison, can hinder productivity and user experience. By implementing a search function akin to VB6’s, twinBASIC could greatly enhance usability for those familiar with traditional VB development practices.

**Alternatives Considered:** I have tried to adapt to the search functionalities provided in .NET for languages like C++ and C#, and while they may serve well in more complex codebases, they fall short for languages and projects where VB6-like efficiency is expected. The complexity and inconsistency of these newer search tools do not meet the needs addressed by VB6’s straightforward search functionality.

**Additional Context:** It's also worth noting that usability details, such as the focus behavior and interface feedback (like byte counts), currently detract from the functionality of the search feature. These aspects should be refined to ensure that the search tool is not only functional but also user-friendly and efficient.

**Conclusion:** A return to the principles of VB6’s search functionality could significantly improve the developer experience in twinBASIC. Thank you for considering this enhancement to make twinBASIC even more robust and intuitive for all users.

**Comment by yereverluvinunclebert** [comment]
I have used several editors that have docked, modal  or non-modal search functionality. I believe in offering choice but for a VB6 100% language compatible product there are going to some expectations regarding the IDE. During debugging the Search facility is probably one of the most heavily used components of the the old VB6 IDE. Therefore I would agree with Anton that a replication of the current search as found in the VB6 IDE would be very useful indeed to smooth transition from VB6 to TwinBasic.

I use RJTextEd for js development and it allows the choice between a bottom-loaded search facility below the code window or a non-modal pop up. That for me is the best choice.

I am just adding my suggestions for a fully functional search non-modal pop up.

![Twinbasic-modal-search](https://github.com/user-attachments/assets/03f9674e-650e-4ac7-99ab-1ffdbf74c208)

![Twinbasic-search-box](https://github.com/user-attachments/assets/19a5b838-bc6e-48fe-8a62-8e083bdf5465)

These are just two suggestions for the possible layout for a simple FIND pop-up based upon the previous functionality but with some more advanced options. The buttons... they are big and clear and do what they say. That is a big bonus.

My penn'orth re: the current search functionality in the TB IDE. I really dislike the tiny FIND facility tucked up in the corner, the whole thing, icons included, too small, not intuitive for a newbie to coding. It might be familiar to a vsCode user but I don't care about them. However, it should be provided as an option. Don't take it away.

Finally my thoughts on the Wayne's enhanced global search, I like it. I don't like it when it first pops up as a floating window as it feels rather useless there. It comes into its own when docked horizontally at the bottom above the debug console, diags &c. It should be there by default. I think it should remain coupled with something like what I have suggested above. The benefit of a pop up is that it has a lot of space to allow future improved functionality as useful options such as those shown.

**Comment by fafalone** [comment]
I think the default find box should work just like VB6 plus enhancements, but I also really like the global search addin. We definitely need to keep that or something like it that shows all results in a list; I didn't care for the VB6 way of having to click next dozens of times in a project-wide search.

I like most of the additional options the screenshots above except the idea of symbols taking on new meaning... At the very least just commandeer a single escape sequence not common in VB/tB code because id definitely want $ to search for $.

**Comment by yereverluvinunclebert** [comment]
> I think the default find box should work just like VB6 plus enhancements, but I also really like the global search addin. We definitely need to keep that or something like it that shows all results in a list; I didn't care for the VB6 way of having to click next dozens of times in a project-wide search.

Agreed. The option to list all items found and display within Wayne's glob search would be my preferred. I don't mind the VB6 way of having to click through each found occurrence as long as there are only 2-3 but when there 100s it annoys.

> I like most of the additional options the screenshots above except the idea of symbols taking on new meaning... At the very least just commandeer a single escape sequence not common in VB/tB code because id definitely want $ to search for $.

The example number 3 above was just to show some sort of potential enhancement clickable through an icon on the form in the context of the search string box. I have something similar in other editors and it is a good idea to take note of what others are doing.

**Comment by yereverluvinunclebert** [comment]
created a small VB project with my proposed non-modal search form and new options implemented as controls only, not yet functional but there to give the idea some flesh.

![find0001](https://github.com/user-attachments/assets/ac890913-7a30-4264-b3a8-53bebaa9736a)

[find.zip](https://github.com/user-attachments/files/18233835/find.zip)

**Comment by yereverluvinunclebert** [comment]
![vb6Find](https://github.com/user-attachments/assets/3516ac17-f227-4073-b1bb-93a8ab1350cd)

**Comment by yereverluvinunclebert** [comment]
As the code editor is based upon the Monaco editor it may be a little more difficult to implement to integrate my proposed changes, I suppose I could look at Wayne's  Global Search and dig into that...

**Comment by FullValueRider** [comment]
As someone who recently created himself a bit of  a headache after a number of search and replaces within a module I'd caution against a global replace.  What we need is to do what Rubberduck did and allow replacement of tokens, not text.  By all means allow a global search to find things, but a global replace that is only text based........

**Comment by fafalone** [comment]
Should have both options.

**Comment by yereverluvinunclebert** [comment]
That is the danger of a find all option. Absolute power corrupts absolutely.  

In my suggestion above, the Find ALL activates only when List ALL or hightlight ALL radio buttons are checked. I have similar find all functionality in my other editors, you just have to know precisely what you are doing before you take this sort of action. Also, you must have decent source control in place so you can revert any recent changes easily.

**Comment by bclothier** [comment]
I would encourage to take a look at how MZ-Tools implement their Find and Replace functionality. I think it's quite brilliant how it allows one to find everything, then be able to edit the search results before doing any replaces. I find it much easier to do a global find, get 100 results, delete 3 or 4 false positive results from the results window, then do a replace on the remaining 96-97 results. That is far much easier to achieve than having to write a regular expression or complex exclusion logic to take out the false positives. 

For those who don't have it, you can see description of its [find dialog here](https://www.mztools.com/v8/onlinehelp/MZTools8Help.html?find_replace_text.htm) and its [result windows here](https://www.mztools.com/v8/onlinehelp/MZTools8Help.html?find_replace_text.htm). 

That said, I agree a token-based search will be much preferable to a text-based search, particularly for refactoring code. However, I still find myself using text-based searches when I'm not necessarily doing refactoring (e.g. reviewing/editing in-line documentation lends itself to the text-based search more better than token-based search). 

I'm also not a big fan of popup window. I find them very annoying because they have to be moved around, and it gets moved around when running searches particularly when the matching text would be obscured by the popup window. For that reason, I tend to prefer them to be docked so they stay where they are and I can multitask more efficiently. YMMV, of course.

**Comment by yereverluvinunclebert** [comment]
The pop up window should be an option, obviously for continuity's sake for VB users. The reason we have the piddlingly small and annoying (IMO) find utility in the top right of the editor is only because we are inheriting the Monaco method fo doing things. I am not a big fan. So let's just put that to bed and giove everyone what they want. RJTextEd manages both.

I have MZTools 3 and 8 and still use both. It is good. Their search is a model for what we might extend the List ALL search to. However, I still use the plain old VB style search for 95% of what I do. It is quick and simple and find next is enough for most use. 

When I want an advance search I can always use the MZ tools version. Don't want to over complicate things for newbs and VB6ers who might find a typical VB style non-modal pop up easier to operate.

In some of my forms I display the basics, then an advanced dropdown to display the extra options, so we could retain the VB6 style simple search and offer the other options as advanced extras.

**Comment by bclothier** [comment]
I guess I'm confused why we are wanting a popup dialog with multiple options if we want to be able to find quickly. That seems to be at cross purpose. 

If we just want to do a quick search, ctrl + F, type in whatever, hit return to get to the next result, press ESC to keep typing is usually all that's needed. Monaco, Visual Studio and VB6 all work the same way in this regard. When we do need to do anything a bit more complicated, we are already disengaged because at that point, we must either use the mouse or perform elaborate keystrokes. At that point, a full blown search has the same utility as a simplified search and having a "advanced" button would be just an extra click. I don't see a point in having a simple, medium and advanced layouts; either have the simple and advanced layout or use advanced layout only (e.g., use the same layout for all search scenarios). 

The way it is, Monaco's and Visual Studio's Find dialog is quite sufficient for finding the next occurrence without having to play "Where's Waldo" that VB6 Find dialog is so fond of playing. The only options I bother with is the "find in selection" which I should point out that in Monaco is implemented quite poorly in contrast to Visual Studio. In Visual Studio, the find in selection is very easy to grok and use, as it's often more useful to do find/replace all but within only a section of code and can be done from a simple dialog. 

But as soon as I start to worry about matching lettercases, whole words, or need to look somewhere else, I'm already disengaged from the typing and searching is now the primary task in front of me. At that point, the simple dialog, whether we talk about Monaco's, Visual Studio's or VB6's is no longer serviceable and we are looking at a full blown search a la MZ-Tools' or Wayne's global find. 

Just my $0.02ZWD.

**Comment by yereverluvinunclebert** [comment]
Horses for courses and gave my penn'orth earlier, what you like and what I like... it is just an interface right, so the addition of a few items on a form versus another way of doing things... 

![searchMin0001](https://github.com/user-attachments/assets/5b8f482c-fa52-4203-b2c2-75220bf0e883)

In the zip above, I just added a button and menu options to switch from basic to advanced.

**Comment by WaynePhillipsEA** [comment]
BETA 752 adds the initial implementation of a FIND/REPLACE panel.  It can be used floating, or docked as any ordinary panel.

<img width="314" alt="Image" src="https://github.com/user-attachments/assets/1aeb7de6-f95f-4498-897f-62b128230fb4" />

We will first attempt to get something roughly equivalent to the VBx standard Find/Replace dialog before later offering improvements.  This initial version does not support project-scope searching, nor does it support pattern matching.

**Comment by yereverluvinunclebert** [comment]
VERY good. With changes like this it will be very easy to make TB my daily driver.

**Comment by yereverluvinunclebert** [comment]
Gimme that Close button!

**Comment by yereverluvinunclebert** [comment]
Can we also have the default Keypress CTRL+H assigned by default to popping up the search box in replace mode? That is how it was in VB6 and that is now hard-coded into my DNA. I don't know why that key combination was chosen but it is how it is.

I do enjoy the CTRL+R keypress as the standard for modern IDEs but the default for an old VB6-er is CTRL+H, I know I can configure keypresses but for continuities sake it should be the default for the oldies and so I don't have to reconfigure each new version, I keep each TB version in its own folder.

**Comment by WaynePhillipsEA** [comment]
> I know I can configure keypresses but for continuities sake it should be the default for the oldies and so I don't have to reconfigure each new version, I keep each TB version in its own folder.

Just FYI, customized keyboard shortcuts are not persisted in the IDE folder.  They are stored in the registry, and so once defined are available in any version of tB.

**Comment by yereverluvinunclebert** [comment]
I know it sounds ridiculous but that Close button is really quite important. All the controls on that form are easy to find, easy to select and easy to use. Once you are used to that juicy big Cancel/Close button then the piddly little 'x' that Microsoft deigns to allow us to use is in the wrong place and is just far too small. I only use the X button when a form won't die. It is the emergency option.

In normal operation I go straight to close button.

The truth is that in a year or so, MS might make that 'X'  even smaller and harder to use.

The Cancel/close button in its old previous location is well-known to all VB6ers and I find myself looking for it when it is not there. My workflow is interrupted.

You have plenty of space on that form, in every way the form looks and operates like the old VB6 form except in that one respect. 

A nice smooth workflow is what we need. Please, stick it there.

**Comment by WaynePhillipsEA** [comment]
The reason I didn't add it is because it doesn't make sense to have a Close button when the panel is docked, since we don't offer an explicit close button for any other panel.

However, when the panel is floating, I do agree that having that close button would offer consistency with the VBx find/replace dialog.  Perhaps I will hide/show the button depending on whether the panel is floating.

**Comment by yereverluvinunclebert** [comment]
Yes, that makes sense. I had never envisaged using it in any other way than floating in typical VB6 manner. I simply move the search box out of the window and I don't have to worry about the box intruding upon work space. I'd never imagined it could be docked.

PS I have only just realised the search box is currently constrained within the IDE window.

**Comment by KarlSevenSeven** [comment]
I like the MZ-Tools Search dialog:
![Image](https://github.com/user-attachments/assets/78515be8-e982-4822-9e25-0acd164dd207)
The results are in a list:
![Image](https://github.com/user-attachments/assets/c20a05c9-b4de-493b-af29-3edb026b41a6)

A double click jumps to the found place.
Then I can immediately check and edit.
Or replace all.

**Comment by fafalone** [comment]
Wayne has mentioned that's planned, so it can replace the Global Search addin completely.

IMO the layout of Global Search is better than a tree like that.

![Image](https://github.com/user-attachments/assets/6997921a-3349-42c4-a8c6-4a25e3ee375d)

The only features it could use are options to set max results per file and not add multiple entries for the same line.

**Comment by fafalone** [comment]
Another feature the new F/R panel should have: highlighting all matches (at least in the active editor/scope) like the Monaco widget.

Makes multiple matches very easy to see.

**Comment by yereverluvinunclebert** [comment]
Desperate for the scope option "Current Procedure". Searching and replacing in TwinBasic is too dangerous to be trusted without local procedure scope encapsulating my possible mistakes. Still coding in VB6 just because of the security that the scope control gives me, in VB6.

**Comment by wqweto** [comment]
Btw, refactor-rename will eliminate this whole class of errors when used with local vars/params. I found this very useful in Visual Studio and VS Code so use it quite often.

**Comment by yereverluvinunclebert** [comment]
Do we have that here in TB within the editor?
